File: src/tools/recall.ts
import { loadMemory, memoryEnabled, renderMemory } from "../memory.js";
import type { ToolDefinition } from "../types.js";

export const NOTHING_REMEMBERED =
  "Nothing remembered for this project yet. This is a fresh start, not a lost history.";

export default {
  name: "recall",
  description:
    "Return the plan and the notes saved for this project by earlier turns or earlier conversations. Call this when you have lost the thread — a new chat, a task you are resuming, or any point where you are about to ask the user to repeat something they may already have told you. It is cheap and returns nothing when nothing was stored.",
  inputSchema: {
    type: "object",
    properties: {},
    additionalProperties: false,
  },
  outputSchema: {
    type: "object",
    properties: {
      content: { type: "string", description: "The stored plan and notes, or a line saying there are none." },
    },
  },
  handler: async (_args, config) => {
    if (!memoryEnabled(config)) {
      return {
        content: [{
          type: "text",
          text: "Persistent memory is disabled on this server (memory.enabled is false). Nothing is stored between conversations.",
        }],
      };
    }
    const rendered = renderMemory(loadMemory(config));
    return { content: [{ type: "text", text: rendered ?? NOTHING_REMEMBERED }] };
  },
} satisfies ToolDefinition;

File: src/tools/recall.rs
use async_trait::async_trait;
use serde_json::{Value, json};

use crate::exec_sessions::SessionState;
use crate::memory::{load_memory, memory_enabled, render_memory};
use crate::tool::Tool;
use crate::types::{AppConfig, ToolResult};

pub const NOTHING_REMEMBERED: &str =
    "Nothing remembered for this project yet. This is a fresh start, not a lost history.";

pub struct Recall;

#[async_trait]
impl Tool for Recall {
    fn name(&self) -> &'static str {
        "recall"
    }

    fn description(&self) -> String {
        "Return the plan and the notes saved for this project by earlier turns or earlier conversations. Call this when you have lost the thread — a new chat, a task you are resuming, or any point where you are about to ask the user to repeat something they may already have told you. It is cheap and returns nothing when nothing was stored.".into()
    }

    fn input_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {},
            "additionalProperties": false
        })
    }

    fn output_schema(&self) -> Option<Value> {
        Some(json!({
            "type": "object",
            "properties": {
                "content": { "type": "string", "description": "The stored plan and notes, or a line saying there are none." }
            }
        }))
    }

    async fn call(&self, _args: Value, config: &AppConfig, _session: &SessionState) -> ToolResult {
        if !memory_enabled(config) {
            return ToolResult::text(
                "Persistent memory is disabled on this server (memory.enabled is false). Nothing is stored between conversations.",
            );
        }

        let config = config.clone();
        let memory = match tokio::task::spawn_blocking(move || load_memory(&config)).await {
            Ok(m) => m,
            Err(e) => return ToolResult::error(e.to_string()),
        };

        let rendered = render_memory(&memory).unwrap_or_else(|| NOTHING_REMEMBERED.to_string());
        ToolResult::text(rendered)
    }
}
